### PR TITLE
feat: add metrics metadata for auth mode and custom operations

### PR DESCRIPTION
**Pull request by dpilch**
#### Description of changes

Add additional metrics metadata for `authorizationMode` and `customOperations`. The metric data is stored in the CFN stack description.

```typescript
{
  "metadata": {
    "dataSources": "dynamodb,mysql,postgres", // exists today

    // new fields below
    /*
      Possible options
      'aws_iam',
      'openid_connect',
      'amazon_cognito_identity_pools',
      'amazon_congnito_user_pools',
      'api_key',
      'aws_lambda',
    */
    "authorizationModes", "api_key,aws_lambda",
    "customOperations": "queries,mutations"
  }
}
```

##### CDK / CloudFormation Parameters Changed

The metrics are stored in the CFN description. No real functional change, but the descriptions will be updated.

#### Issue #, if available

N/A

#### Description of how you validated changes

* Unit tests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-api-construct/src/__tests__/__functional__/metadata.test.ts [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+          lambdaConfig: {
+            function: new NodejsFunction(stack, 'TestAuthorizer', { entry: path.join(__dirname, 'authorizer.ts') }),
+            ttl: Duration.days(7),
+          },
+        },
+      });
+      const template = Template.fromStack(stack);
+      expect(JSON.parse(template.toJSON().Description).metadata).toMatchInlineSnapshot(`
+        Object {
+          "authorizationModes": "api_key,aws_iam,aws_lambda",
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct/src/__tests__/__functional__/metadata.test.ts:354]: This asserts order as well as content -- is that intentional?

File: packages/amplify-graphql-api-construct/src/internal/metadata.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+export const getMetadataAuthorizationModes = (authorizationModes: AuthorizationModes): string => {
+  const configKeyToAuthMode: Record<string, string> = {
+    iamConfig: 'aws_iam',
+    oidcConfig: 'openid_connect',
+    identityPoolConfig: 'amazon_cognito_identity_pools',
+    userPoolConfig: 'amazon_cognito_user_pools',
+    apiKeyConfig: 'api_key',
+    lambdaConfig: 'aws_lambda',
+  };
+  const authModes = Object.keys(authorizationModes)
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct/src/internal/metadata.ts:19]: One of the tests above asserts the order of the auth mode metric. Is that intentional? Should we sort the values here so we can provide a stable value irrespective of the order in which they are declared?

File: packages/amplify-graphql-api-construct/src/internal/metadata.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    .map((mode) => configKeyToAuthMode[mode])
+    // remove values not found in mapping
+    .filter((mode) => !!mode);
+  return authModes.join(',');
+};
+
+export const getMetadataCustomOperations = (definition: IAmplifyGraphqlDefinition): string => {
+  const customOperations: string[] = [];
+  if (definition.schema.includes('type Query')) {
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct/src/internal/metadata.ts:28]: Why not subscriptions?

**Comment by palpatim** [comment]
As far as I can tell this is a typo in the PR description, but calling it out in case:

```
    /*
      Possible options
      'aws_iam',
      'openid_connect',
      'amazon_cognito_identity_pools',
      'amazon_congnit_user_pools',
      'api_key',
      'aws_lambda',
    */
```

File: packages/amplify-graphql-api-construct/src/internal/metadata.ts (hunk 3 above)
Comment (dpilch) [reply packages/amplify-graphql-api-construct/src/internal/metadata.ts:28]: Wasn't in the product requirements, but we should probably just add it.

File: packages/amplify-graphql-api-construct/src/internal/metadata.ts (hunk 2 above)
Comment (dpilch) [reply packages/amplify-graphql-api-construct/src/internal/metadata.ts:19]: Good idea, I will apply a sort.

**Comment by dpilch** [comment]
Yes, typo in the description. We can see `amazon_congnito_user_pools` is set correctly in the unit test.

**Review by palpatim — APPROVED** [review]

**Review by atierian — APPROVED** [review]

**Pull request merged**